### PR TITLE
Implements the Tesla engine and supporting features

### PR DESCRIPTION
**Pull request by Leshana**
## Tesla Engine
Ported from Paradise and TG; many authors involved
image
### Items added
- Adds the Tesla energy ball.  Its created by bombarding the Tesla generator with the Particle Accelerator.  Its a big ball that shoots lightning bolts at stuff.   Don't let it hit you.   Luckily the singularity containment field works on the energy ball too.
- Adds the Tesla coil and grounding rod machines.  They are how you actually gather power from the energy ball and protect the station from lightning bolts.
   - They are both constructable in game, get circuits from RnD and autolathe respectively.
- Adds the Tesla generator. This is required to actually make the Tesla energy ball.  **This is not mapped in** and the generator can't be built or ordered from cargo.  Therefore players can't actually get the Tesla engine running in game unless an admin spawns it or it is added to the map.
- In game manual forthcoming in case its decided to let players build it.

### Supporting technologies
- Ports typecache list procs from TG.  These are a fast way to filter a list of atoms by type.
- Ports the "orbit" feature and subsystem from TG ( tgstation/tgstation#10960 tgstation/tgstation#20632 )
- Adds utility methods for common machinery behavior.  I have wanted this for awhile. It could get even more sophisticated if we wanted, but `default_unfasten_wrench` takes care of the most common repeated code in machinery `attackby`

### CODE REVIEW

**Comment by Leshana** [comment]
@Atermonera fixed build (test) issue.

**Comment by Arokha** [comment]
How 'bout that tesla, @Anewbe

**Comment by Anewbe** [comment]
Oh right! That!

File: code/modules/power/tesla/energy_ball.dm [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+										/obj/structure/particle_accelerator/particle_emitter/left,
+										/obj/structure/particle_accelerator/particle_emitter/right,
+										/obj/structure/particle_accelerator/power_box,
+										/obj/structure/particle_accelerator/end_cap,
+										/obj/machinery/containment_field,
+										/obj/structure/disposalpipe,
+										/obj/structure/sign,
+										/obj/machinery/gateway,
+										/obj/structure/lattice,
+										/obj/structure/grille,
```
Comment (Cyantime) [inline code/modules/power/tesla/energy_ball.dm]: Catwalks too, along the same reasoning as lattice and grilles?

**Pull request merged**